File: SliderViewHorizontalIndicator.swift
//
//  SliderIndicator.swift
//  https://github.com/ysoftware
//
//  Copyright © 2017 Yaroslav Erohin. All rights reserved.
//

import UIKit

@IBDesignable
class SliderViewHorizontalIndicator: UIView, SliderViewIndicator {

    // MARK: - Properties

    /** Lowest value of the indicator. Cannot be higher than `maxValue`. Default value is 0.0. */
    @IBInspectable var minValue:Float = 0.0 {
        didSet {
            if minValue >= maxValue { minValue = maxValue - 0.1 }
            reloadViews()
        }
    }

    /** Highest value of the indicator. Cannot be lower than `minValue`. Default value is 1.0. */
    @IBInspectable var maxValue:Float = 1.0 {
        didSet {
            if minValue >= maxValue { maxValue = minValue + 0.1 }
            reloadViews()
        }
    }

    /** Current value of the indicator. Default value is `0.5`. */
    @IBInspectable var value:Float = 0.5 {
        didSet {
            if value > maxValue { value = maxValue }
            else if value < minValue { value = minValue }
            reloadViews()
        }
    }

    /** Relative padding of the indicator. From `0.0` to `1.0`. Default value is `0`. */
    @IBInspectable var relativePadding:Float = 0 {
        didSet {
            if relativePadding >= 1 { relativePadding = 0.99 }
            else if relativePadding < 0 { relativePadding = 0 }
            reloadViews()
        }
    }

    /** Colour of the indicator. Default value is `black`.

     To set background colour, use property of `UIView` called `backgroundColor`. */
    @IBInspectable var foregroundColour: UIColor = .white {
        didSet {
            foregroundView.backgroundColor = foregroundColour
        }
    }

    /** Defines the behaviour of the indicator. Default value is `center`. */
    var anchor:Anchor = .center { didSet { reloadViews() }}

    /** Behaviour of the SliderIndicator. */
    enum Anchor {
        /** Goes from left side all the way to the right. */
        case left
        /** Goes from right side all the way to the left. */
        case right
        /** Middle point is at the center, `minValue` on the left, `maxValue` is on the right side.  */
        case center
    }

    /** Middle point between `minValue` and `maxValue`. */
    var middleValue:Float {
        return (minValue + maxValue) / 2
    }

    // MARK: - Private

    private let foregroundView = UIView()

    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(foregroundView)
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        addSubview(foregroundView)
    }

    override func layoutSubviews() {
        super.layoutSubviews()

        reloadViews()
    }

    private func reloadViews() {
        UIView.animate(withDuration: 0.1,
                       delay: 0,
                       usingSpringWithDamping: 0.8,
                       initialSpringVelocity: 0,
                       options: .curveEaseOut,
                       animations: {
                        self.foregroundView.backgroundColor = self.foregroundColour

                        let dif = self.maxValue - self.minValue
                        let width = Float(self.frame.size.width)

                        switch self.anchor {
                        case .center:
                            let isPositive = self.middleValue - self.value < 0
                            let relativeValue = self.middleValue - self.value
                            let center = width / 2
                            let indicatorWidth = center - (center * self.relativePadding)
                            let position = abs(indicatorWidth / dif * 2 * relativeValue)

                            self.foregroundView.frame = CGRect(
                                x: CGFloat(isPositive ? center : center - position),
                                y: 0,
                                width: CGFloat(position),
                                height: self.frame.size.height)
                        case .left, .right:
                            let indicatorWidth = width - (width * self.relativePadding)
                            let position = indicatorWidth * (self.value - self.minValue)
                                / (self.maxValue - self.minValue)

                            self.foregroundView.frame = CGRect(
                                x: self.anchor == .left ? 0 : CGFloat(width - position),
                                y: 0,
                                width: CGFloat(position),
                                height: self.frame.size.height)
                        }
        }, completion: nil)
    }
    
}

File: SliderViewHorizontalIndicator.java
package ru.firstpro.photoeditor;

import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;
import android.widget.RelativeLayout;

import static ru.firstpro.photoeditor.SliderViewHorizontalIndicator.Anchor.left;

/**
 * Created by ysoftware on 24.05.17.
 */

public class SliderViewHorizontalIndicator extends RelativeLayout {
	public SliderViewHorizontalIndicator(Context context, AttributeSet attributeSet) {
		super(context, attributeSet);
		foregroundView = new RelativeLayout(context);
		addView(foregroundView);
	}

	/* VALUES */

	public void setMaxValue(float v) {
		maxValue = v;
		if (minValue >= maxValue) { maxValue = minValue + 0.1f; }
		reloadViews();
	}

	public void setMinValue(float v) {
		minValue = v;
		if (minValue >= maxValue) { minValue = maxValue - 0.1f; }
		reloadViews();
	}

	public void setValue(float v) {
		value = v;
		if (value > maxValue) { value = maxValue; }
		else if (value < minValue) { value = minValue; }
		reloadViews();
	}

	public float getValue() {
		return value;
	}

	public float getMinValue() {
		return minValue;
	}

	public float getMaxValue() {
		return maxValue;
	}

	public float getMiddleValue() {
		return (minValue + maxValue) / 2;
	}

	/* SETTINGS */

	enum Anchor { left, right, center }

	public void setAnchor(Anchor a) {
		anchor = a;
		reloadViews();
	}

	public void setRelativePadding(float v) {
		relativePadding = v;
		if (relativePadding >= 1) { relativePadding = 0.99f; }
		else if (relativePadding < 0) { relativePadding = 0f; }
		reloadViews();
	}

	public void setForegroundColor(int c) {
		foregroundColor = c;
		foregroundView.setBackgroundColor(foregroundColor);
	}

	public Anchor getAnchor() {
		return anchor;
	}

	public float getRelativePadding() {
		return relativePadding;
	}

	public int getForegroundColor() {
		return foregroundColor;
	}

	/* PRIVATE */

	private float maxValue = 1f;
	private float minValue = 0f;
	private float value = 0.5f;
	private float relativePadding = 0f;
	private int foregroundColor = Color.parseColor("#FFF25F"); // a nice yellow color by default
	private Anchor anchor = left;
	private RelativeLayout foregroundView;

	void reloadViews() {
		foregroundView.setBackgroundColor(foregroundColor);

		float dif = maxValue - minValue;
		float width = getMeasuredWidth();
		int height = getMeasuredHeight();

		switch (anchor) {
			case center: {
				Boolean isPositive = getMiddleValue() - value < 0;
				float relativeValue = getMiddleValue() - value;
				float center = width / 2;
				float indicatorWidth = center - (center * relativePadding);
				float position = Math.abs(indicatorWidth / dif * 2 * relativeValue);

				int _width = Math.round(position);
				int _x = Math.round(isPositive ? center : center - position);

				LayoutParams params = new LayoutParams(_width, height);
				params.setMarginStart(_x);
				foregroundView.setLayoutParams(params);
				break;
			}
			case left:
			case right: {
				float indicatorWidth = width - (width * relativePadding);
				float position = indicatorWidth * (value - minValue) / (maxValue - minValue);

				int _width = Math.round(position);
				int _x = anchor == left ? 0 : Math.round(width - position);

				LayoutParams params = new LayoutParams(_width, height);
				params.setMarginStart(_x);
				foregroundView.setLayoutParams(params);
				break;
			}
		}
	}
}